api/v0/ingest/models/discover_request.go

```diff
@@ -0,0 +1,80 @@
+package models
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/filecoin-project/storetheindex/config"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p-core/record"
+)
+
+type DiscoverRequest struct {
+	ProviderID    peer.ID
+	DiscoveryAddr string
+	Seq           uint64
+}
+
+// DiscoverRequestEnvelopeDomain is the domain string used for discover requests contained in a Envelope
+const DiscoverRequestEnvelopeDomain = "indexer-discover-request-record"
+
+// DiscoverRequestEnvelopePayloadType is the type hint used to identify DiscoverRequest records in a Envelope.
+var DiscoverRequestEnvelopePayloadType = []byte("indexer-discover-request")
+
+func init() {
+	record.RegisterType(&DiscoverRequest{})
+}
+
+// Domain is used when signing and validating DiscoverRequest records contained in Envelopes
+func (r *DiscoverRequest) Domain() string {
+	return DiscoverRequestEnvelopeDomain
+}
+
+// Codec is a binary identifier for the DiscoverRequest type
+func (r *DiscoverRequest) Codec() []byte {
+	return DiscoverRequestEnvelopePayloadType
+}
+
+// UnmarshalRecord parses a DiscoverRequest from a byte slice
+func (r *DiscoverRequest) UnmarshalRecord(data []byte) error {
+	if r == nil {
+		return fmt.Errorf("cannot unmarshal DiscoverRequest to nil receiver")
+	}
+
+	return json.Unmarshal(data, r)
+}
+
+// MarshalRecord serializes an DiscoverRequesr to a byte slice.
+func (r *DiscoverRequest) MarshalRecord() ([]byte, error) {
+	return json.Marshal(r)
+}
+
+// MakeDiscoverRequest creates a signed DiscoverRequest and marshals it into bytes
+func MakeDiscoverRequest(providerIdent config.Identity, discoveryAddr string) ([]byte, error) {
+	providerID, privKey, err := providerIdent.Decode()
+	if err != nil {
+		return nil, err
+	}
+
+	req := &DiscoverRequest{
+		ProviderID:    providerID,
+		DiscoveryAddr: discoveryAddr,
+		Seq:           peer.TimestampSeq(),
+	}
+
+	return makeRequestEnvelop(req, privKey)
+}
+
+// ReadDiscoverRequest unmarshals a DiscoverRequest from bytes, verifies the
+// signature, and returns the DiscoverRequest
+func ReadDiscoverRequest(data []byte) (*DiscoverRequest, error) {
+	_, untypedRecord, err := record.ConsumeEnvelope(data, DiscoverRequestEnvelopeDomain)
+	if err != nil {
+		return nil, fmt.Errorf("cannot consume discover request envelope: %s", err)
+	}
+	rec, ok := untypedRecord.(*DiscoverRequest)
```

Comment:
Out of a curiosity: we check here that the signature in the record is valid. Do we need to check also the source (i.e. that the signer is the right one?). The same applies to the other records.

Reply:
I think we have to rely on the source keeping their private key private.  The source's ID is the public key that is tied to that private key, and that peer ID is how we identify any source.

That being said, perhaps we need a way for a source to change their ID in case the think their private key has been compromized.  The current mechanism for that would be to do `DeleteProvider` and then do discovery using the new ID.